### ISSUE TITLE
Implement CalibrationAnalysis

### ISSUE THREAD

**Issue by neubig** [neulab/ExplainaBoard#417]
Calibration is whether a system's confidence is well-correlated with whether the system got the answer right or not. It would be nice if we could do analyses related to calibration, such as calculating expected calibration error: https://arxiv.org/abs/1706.04599

I think this should probably be implemented as an additional variety of analysis, which would be simple and self-contained: https://github.com/neulab/ExplainaBoard/blob/main/explainaboard/analysis/analyses.py#L45

**Comment by pfliu-nlp** [comment]
@neubig, this is also one feature I have been expecting.  The only complicated thing is we need `probability` as an additional feature.
We probably can have some rules like this: if a system output file from a classification task contains the feature of `probability`, then the processor will conduct calibration adaptively, (kinda similar to training set dependent features)

**Comment by neubig** [comment]
At the moment, if an analysis is not applicable it returns `None`, so we could do a similar thing here.

**Comment by pfliu-nlp** [comment]
Yes, I noticed that. But this also will result in potential bugs when deploying the web platforms, which I have debugged for a pretty long time.
It's a potential schema validation bug that will happen here: https://github.com/neulab/explainaboard_web/blob/c711cf8277c4f19d0c12e18b008b7ec2b8779d00/backend/src/impl/default_controllers_impl.py#L447

**Comment by neubig** [comment]
I don't think returning `None` is necessarily a bad thing if we know it's expected behavior. But we could definitely discuss ways to rectify this if it's a problem.

**Comment by odashi** [comment]
I just created #418 around the topic of `None`.

Basically I think `None` is not informative as it provides no fine-grained information, and users have no control in case of invalid operations. I prefer either:

- raising appropriate exceptions to inform appropriate information outside the process.
- Using the `Result` semantics (adopted in some other languages, such as [Rust](https://doc.rust-lang.org/std/result/enum.Result.html)). the result type have either the data or error as data.

**Comment by odashi** [comment]
I think it is better that `None` is used to inform only *no information*. If there are some information that are useful to inform externally, it is good to adopt either way described above.

**Comment by qjiang002** [comment]
Here are some draft ideas of implementing calibration analysis:
* When to perform calibration analysis: (1) classification tasks: text_classification, text_pair_classification, aspect_based_sentiment_classification, tabular_classification; (2) users provide both predicted labels and confidence values in the output file. Check that users provide confidence values in range [0, 1], not logits. 
* Divide the samples into K bins where K is a hyper-parameter. We can either divide confidence range [0, 1] into K bins, or divide samples evenly into K bins. For now we choose to divide the confidence range. 
* Calculate the accuracy and average confidence of each bin, and calculate ECE and MCE according to formula (3) and (5) in this [paper](https://arxiv.org/pdf/1706.04599.pdf)

**Comment by neubig** [comment]
One comment: we may want to view a `CalibrationAnalysisResult` as either a subclass of a `BucketAnalysisResult`, or a `BucketAnalysisResult` with some auxiliary information (namely ECE and MCE). That would make it easy to, for example, display the calibration diagram using (nearly) the same code that we normally use for displaying buckets.

**Comment by odashi** [comment]
@neubig Subclassing should be used when they satisfy the is-a relationship. If we need only the same functionality, implementing separate code is better to organize the whole thing.

**Comment by odashi** [comment]
This issue was resolved.
